src/System.Management.Automation/engine/MshObject.cs

```diff
@@ -2347,6 +2378,204 @@ internal bool IsHelpObject
         #endregion
     }
 
+    /// <summary>
+    /// A factory class for more efficient creation of PSCustomObjects
+    /// </summary>
+    /// <para/>
+    /// This creates a context in which the guarantee is given that the
+    /// object being created, and the members that are being added are not
+    /// visible from any other location.
+    /// This way, defensive copies and otherwise expensive operations can be avoided.
+    public ref struct PSObjectBuilder
+    {
+        const int DefaultHirearchyDepth = 4;
+        private readonly HashSet<string> _validatedNames;
+        private Stack<PSObject> _currentObjectStack;
+
+        private PSObject _currentObject;
+
+        private Stack<PSObject> CurrentObjectStack => _currentObjectStack ?? (_currentObjectStack = new Stack<PSObject>(DefaultHirearchyDepth));
```

Comment:
Ah, we could implement ValueStack struct to allocate init segment in stack. (I see ValueStringBuilder is used internally in CoreFX.) Seems we could exclude tons short-time allocations. The same is true for HashSet ( _validatedNames).